probe/endpoint/connection_tracker.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/weaveworks/scope/probe/process"
+	"github.com/weaveworks/scope/report"
+)
+
+// connectionTrackerConfig are the config options for the endpoint tracker.
+type connectionTrackerConfig struct {
+	HostID       string
+	HostName     string
+	SpyProcs     bool
+	UseConntrack bool
+	WalkProc     bool
+	UseEbpfConn  bool
+	ProcRoot     string
+	BufferSize   int
+	Scanner      procspy.ConnectionScanner
+	DNSSnooper   *DNSSnooper
+}
+
+type connectionTracker struct {
+	conf            connectionTrackerConfig
+	flowWalker      flowWalker // Interface
+	ebpfTracker     eventTracker
+	reverseResolver *reverseResolver
+}
+
+func newConnectionTracker(conf connectionTrackerConfig) connectionTracker {
+	if !conf.UseEbpfConn {
+		// ebpf OFF, use flowWalker
+		return connectionTracker{
+			conf:            conf,
+			flowWalker:      newConntrackFlowWalker(conf.UseConntrack, conf.ProcRoot, conf.BufferSize, "--any-nat"),
+			ebpfTracker:     nil,
+			reverseResolver: newReverseResolver(),
+		}
+	}
+	// When ebpf will be active by default, check if it starts correctly otherwise fallback to flowWalk
+	et, err := newEbpfTracker(conf.UseEbpfConn)
+	if err != nil {
+		// TODO: fallback to flowWalker, when ebpf is enabled by default
+		log.Errorf("Error setting up the ebpfTracker, connections will not be reported: %s", err)
+		noopConnectionTracker := connectionTracker{
+			conf:            conf,
+			flowWalker:      nil,
+			ebpfTracker:     nil,
+			reverseResolver: nil,
+		}
+		return noopConnectionTracker
+	}
+	// Run conntrack and proc parsing synchronously only once to initialize ebpfTracker
+	seenTuples := map[string]fourTuple{}
+	// Consult the flowWalker to get the initial state
+	if !conf.UseConntrack {
+		log.Warn("conntrack is disabled")
+	} else if err = IsConntrackSupported(conf.ProcRoot); err != nil {
+		log.Warnf("Not using conntrack: not supported by the kernel: %s", err)
+	} else {
+		existingFlows, err := existingConnections([]string{"--any-nat"})
+		if err != nil {
+			log.Errorf("conntrack existingConnections error: %v", err)
+		} else {
+			for _, f := range existingFlows {
+				tuple := fourTuple{
+					f.Original.Layer3.SrcIP,
+					f.Original.Layer3.DstIP,
+					uint16(f.Original.Layer4.SrcPort),
+					uint16(f.Original.Layer4.DstPort),
+				}
+				// Handle DNAT-ed connections in the initial state
+				if f.Original.Layer3.DstIP != f.Reply.Layer3.SrcIP {
+					tuple = fourTuple{
+						f.Reply.Layer3.DstIP,
+						f.Reply.Layer3.SrcIP,
+						uint16(f.Reply.Layer4.DstPort),
+						uint16(f.Reply.Layer4.SrcPort),
+					}
+				}
+				seenTuples[tuple.key()] = tuple
+			}
+		}
+	}
+
+	var processCache *process.CachingWalker
+	var scanner procspy.ConnectionScanner
+	processCache = process.NewCachingWalker(process.NewWalker(conf.ProcRoot))
+	processCache.Tick()
+	scanner = procspy.NewSyncConnectionScanner(processCache)
+	defer scanner.Stop()
+	conns, err := scanner.Connections(conf.SpyProcs)
```

Comment:
This is executed only once as part of the ebpf tracker initialization, so this will affect only the startup time, we will not execute scanner anymore when using ebpf.